### PR TITLE
feat(terraform): Adding yaml based build time policies for corresponding PC runtime policies

### PR DESCRIPTION
**Pull request by praveen-panw**
**By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.**

This PR has 6 YAML-based checks with detailed description below

### 1. Checkov Title - Ensure Container Instance is configured with managed identity

<ins>PC Policy ID</ins> - a9b38d9b-a4df-4f60-bf87-c24d105d2bd9

<ins>PC Policy Title</ins> - Azure Container Instance not configured with the managed identity 

<ins>Compliance standards </ins> - 

`New Zealand Information Security Manual (NZISM v3.4)` `MITRE ATT&CK v10.0` `CIS Controls v7.1` `CIS Controls v8` `PCI DSS v4.0` `Australian Energy Sector Cyber Security Framework (AESCSF)` `ACSC Information Security Manual (ISM)` `Copy of ACSC Information Security Manual (ISM)` `MLPS 2.0 (Level 2)` `Copy 1 0 of ACSC Information Security Manual (ISM)` `Copy 11 of ACSC Information Security Manual (ISM)` `Secure Controls Framework (SCF) - 2022.2.1` `Sarbanes Oxley Act (SOX)`

<ins>Remediation:</ins>

To enable managed identity on your container instance; follow below URL:
https://docs.microsoft.com/en-us/azure/container-instances/container-instances-managed-identity

---

### 2. Checkov Title - Ensure AKS cluster has Azure CNI networking enabled

<ins>PC Policy ID</ins> - ac313c08-1f79-4e55-96e4-49c20064bff0

<ins>PC Policy Title</ins> - Azure AKS cluster Azure CNI networking not enabled

<ins>Compliance standards </ins> - 

`New Zealand Information Security Manual (NZISM v3.4)` `MITRE ATT&CK v10.0` `CIS Controls v7.1` `CIS Controls v8` `PCI DSS v4.0` `Australian Energy Sector Cyber Security Framework (AESCSF)` `ACSC Information Security Manual (ISM)` `Copy of ACSC Information Security Manual (ISM)` `MLPS 2.0 (Level 2)` `Copy 1 0 of ACSC Information Security Manual (ISM)` `Copy 11 of ACSC Information Security Manual (ISM)` `Secure Controls Framework (SCF) - 2022.2.1` `Sarbanes Oxley Act (SOX)`

<ins>Remediation:</ins>

To create a new AKS cluster with the Azure CNI network plugin enabled, follow the below URL:
https://learn.microsoft.com/en-us/azure/aks/configure-azure-cni

---

### 3. Checkov Title - Ensure Azure Container Registry (ACR) has HTTPS enabled for webhook

<ins>PC Policy ID</ins> - 9a28b0fb-67cd-4de9-80b0-702bc0ca6177

<ins>PC Policy Title</ins> - Azure ACR HTTPS not enabled for webhook

<ins>Compliance standards </ins> - 

`NIST 800-53 Rev 5` `CyberSecurity Law of the People's Republic of China` `NIST 800-53 Rev4` `APRA (CPS 234) Information Security` `Cybersecurity Maturity Model Certification (CMMC) v.1.02` `Brazilian Data Protection Law (LGPD)` `NIST SP 800-171 Revision 2` `PCI DSS v3.2.1` `NIST SP 800-172` `HITRUST v.9.4.2` `NIST CSF` `CSA CCM v.4.0.1` `NIST CSF (中文)` `ISO/IEC 27002:2013` `Fedramp (Moderate)` `ISO/IEC 27017:2015` `New Zealand Information Security Manual (NZISM v3.4)` `ISO/IEC 27018:2019` `CIS Controls v7.1` `CIS Controls v8` `FFIEC` `PCI DSS v4.0` `ACSC Information Security Manual (ISM)` `Copy of ACSC Information Security Manual (ISM)` `MLPS 2.0 (Level 2)` `Copy 1 0 of ACSC Information Security Manual (ISM)` `Copy 11 of ACSC Information Security Manual (ISM)` `Secure Controls Framework (SCF) - 2022.2.1` `Sarbanes Oxley Act (SOX)`

<ins>Remediation:</ins>

Update your container registry webhook URI to use HTTPS.

1. Sign in to the Azure portal.
2. Navigate to the container registry in which you want to modify the webhook.
3. Under Services, select Webhooks.
4. Select your existing webhook.
5. Near the top of the next window pane, select Configure.
6. Under Service URI in the next window, modify your URI to use https:// and click Save.

---

### 4. Checkov Title - Ensure VNET subnet is configured with a Network Security Group (NSG)

<ins>PC Policy ID</ins> - d3ed9388-fa76-44b7-ac6f-72503b6340e0

<ins>PC Policy Title</ins> - Azure Virtual Network subnet is not configured with a Network Security Group

<ins>Compliance standards </ins> - 

`PIPEDA` `CCPA 2018` `NIST 800-53 Rev 5` `CyberSecurity Law of the People's Republic of China` `Copy of CCPA 2018-0417` `NIST 800-53 Rev4` `APRA (CPS 234) Information Security` `Cybersecurity Maturity Model Certification (CMMC) v.1.02` `Brazilian Data Protection Law (LGPD)` `NIST SP 800-171 Revision 2` `PCI DSS v3.2.1` `NIST SP 800-172` `HITRUST v.9.4.2` `NIST CSF` `CSA CCM v.4.0.1` `NIST CSF (中文)` `ISO/IEC 27002:2013` `New Zealand Information Security Manual (NZISM v3.4)` `ISO/IEC 27017:2015` `ISO/IEC 27018:2019` `Copy of (AESCSF)-IR` `CIS Controls v7.1` `CIS Controls v8` `PCI DSS v4.0` `ACSC Information Security Manual (ISM)` `Copy of ACSC Information Security Manual (ISM)` `MLPS 2.0 (Level 2)` `Copy 1 0 of ACSC Information Security Manual (ISM)` `Copy 11 of ACSC Information Security Manual (ISM)` `Secure Controls Framework (SCF) - 2022.2.1` `Sarbanes Oxley Act (SOX)`

<ins>Remediation:</ins>

1. Log in to the Azure Portal.
2. Select 'Virtual Networks', and select the virtual network you need to modify.
3. Select 'Subnets', and select the subnet you need to modify.
4. Select the Network security group (NSG) you want to associate with the subnet.
5. 'Save' your changes.

---

### 5. Checkov Title - Ensure private endpoint is configured to key vault

<ins>PC Policy ID</ins> - a03ae1cb-c130-46cb-97ae-02cceab10b07

<ins>PC Policy Title</ins> - Azure Key vault Private endpoint connection is not configured 

<ins>Compliance standards </ins> - 

`New Zealand Information Security Manual (NZISM v3.4)` `CIS Controls v7.1` `CIS Controls v8` `ACSC Information Security Manual (ISM)` `Copy of ACSC Information Security Manual (ISM)` `MLPS 2.0 (Level 2)` `Copy 1 0 of ACSC Information Security Manual (ISM)` `Copy 11 of ACSC Information Security Manual (ISM)` `Secure Controls Framework (SCF) - 2022.2.1` `Sarbanes Oxley Act (SOX)`

<ins>Remediation:</ins>

Refer following URL for configuring Private endpoints on your Key vaults:
https://docs.microsoft.com/en-us/azure/key-vault/general/private-link-service

NOTE: The Key vault associated with private endpoints should not be allowing access from all networks in Firewalls and virtual networks section, make sure the Selected networks are configured with restrictive Virtual networks access. Otherwise the security provided by private endpoints will not be satisfied.

---

### 6. Checkov Title - Azure Storage account is not configured with private endpoint connection 

<ins>PC Policy ID</ins> - f32919fb-62a7-498a-83dd-e31e60ceda29

<ins>PC Policy Title</ins> - Azure Storage account is not configured with private endpoint connection

<ins>Compliance standards </ins> - 

`CIS v1.5.0 (Azure) - Level 1`

<ins>Remediation:</ins>

Refer following URL for configuring Private endpoints on your Storage account:
https://learn.microsoft.com/en-us/azure/private-link/create-private-endpoint-portal?#create-a-private-endpoint

### CODE REVIEW

File: checkov/terraform/checks/graph_checks/azure/AzureAKSclusterAzureCNIEnabled.yaml [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+      attribute: "network_profile.network_plugin"
+      operator: "exists"
+
+    - cond_type: "attribute"
+      resource_types: "azurerm_kubernetes_cluster"
+      attribute: "network_profile.network_plugin"
+      operator: "within"
+      value: 
+        - "azure"
+        - "Azure"
```
Comment (gruebel) [inline checkov/terraform/checks/graph_checks/azure/AzureAKSclusterAzureCNIEnabled.yaml:20]:
```suggestion
    cond_type: "attribute"
    resource_types: "azurerm_kubernetes_cluster"
    attribute: "network_profile.network_plugin"
    operator: "within"
    value: 
      - "azure"
      - "Azure"
```
I can think you don't need the exist block, this should achieve the same.

File: checkov/terraform/checks/graph_checks/azure/AzureContainerInstanceconfigManagedIdentity.yaml [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+  id: "CKV2_AZURE_28"
+  name: "Ensure Container Instance is configured with managed identity"
+  category: "GENERAL_SECURITY"
+
+definition:
+  and:
+    - cond_type: attribute
+      resource_types: azurerm_container_group
+      attribute: identity.type
+      operator: "exists"
```
Comment (gruebel) [inline checkov/terraform/checks/graph_checks/azure/AzureContainerInstanceconfigManagedIdentity.yaml:11]: same here, not needed, but please test it 😄

File: checkov/terraform/checks/graph_checks/azure/AzureKeyVaultConfigPrivateEndpoint.yaml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+definition:
+  and:
+
+    - cond_type: "filter"
+      attribute: "resource_type"
+      operator: "within"
+      value:
+        - "azurerm_key_vault"
+    - cond_type: "connection"
```
Comment (gruebel) [inline checkov/terraform/checks/graph_checks/azure/AzureKeyVaultConfigPrivateEndpoint.yaml:14]: nice 👍

File: checkov/terraform/checks/graph_checks/azure/AzureSubnetConfigWithNSG.yaml [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      operator: "within"
+      value:
+        - "azurerm_subnet"
+    - cond_type: "connection"
+      resource_types:
+        - "azurerm_subnet"
+      connected_resource_types:
+        - "azurerm_network_security_group"
+        - "azurerm_subnet_network_security_group_association"
+      operator: "exists"
```
Comment (gruebel) [inline checkov/terraform/checks/graph_checks/azure/AzureSubnetConfigWithNSG.yaml:23]:
```suggestion
    - cond_type: "connection"
      resource_types:
        - "azurerm_subnet"
      connected_resource_types:
        - "azurerm_subnet_network_security_group_association"
      operator: "exists"
    - cond_type: "connection"
      resource_types:
        - "azurerm_subnet_network_security_group_association"
      connected_resource_types:
        - "azurerm_network_security_group"
      operator: "exists"
```
shouldn't be like this?

**Review by gruebel — COMMENTED** [review]
nice work, a few improvement suggestions

File: checkov/terraform/checks/graph_checks/azure/AzureContainerInstanceconfigManagedIdentity.yaml (hunk 2 above)
Comment (praveen-panw) [reply checkov/terraform/checks/graph_checks/azure/AzureContainerInstanceconfigManagedIdentity.yaml:11]: Hello @gruebel,

I tested it and this block is mandatory in this policy. It doesn't work without it.

File: checkov/terraform/checks/graph_checks/azure/AzureAKSclusterAzureCNIEnabled.yaml (hunk 1 above)
Comment (praveen-panw) [reply checkov/terraform/checks/graph_checks/azure/AzureAKSclusterAzureCNIEnabled.yaml:20]: Yep, tested. You're right 🙌🏻

I've updated it.

File: checkov/terraform/checks/graph_checks/azure/AzureSubnetConfigWithNSG.yaml (hunk 4 above)
Comment (praveen-panw) [reply checkov/terraform/checks/graph_checks/azure/AzureSubnetConfigWithNSG.yaml:23]: @gruebel, yep we can do that too. I remember you mentioning in the call the other day that, sometimes connections may not work as per intended if clubbed together, but the one which I wrote, I tested it again, it's working fine here. But, if you want, I can modify it as per the suggestion you provided.

Comment (praveen-panw) [reply checkov/terraform/checks/graph_checks/azure/AzureSubnetConfigWithNSG.yaml:23]: Discussed with @gruebel and concluded to keep it as it is 🙂

File: tests/terraform/graph/checks/test_yaml_policies.py [hunk 5]
```diff
@@ -351,6 +351,24 @@ def test_GCPMySQLdbInstancePoint_In_TimeRecoveryBackupIsEnabled(self):
...
+                    self.go("test_AzureACR_HTTPSwebhook")
+
+    def test_AzureSubnetConfigWithNSG(self):
+                    self.go("test_AzureSubnetConfigWithNSG")
+
+    def test_AzureKeyVaultConfigPrivateEndpoint(self):
+                    self.go("test_AzureKeyVaultConfigPrivateEndpoint")
+
+    def test_AzureStorageAccConfigWithPrivateEndpoint(self):
+                    self.go("test_AzureStorageAccConfigWithPrivateEndpoint")
```
[the remark this answers is not in the document]
Comment (gruebel) [reply tests/terraform/graph/checks/test_yaml_policies.py:371]: please double check the folder names

[the remark this answers is not in the document]
Comment (praveen-panw) [reply tests/terraform/graph/checks/test_yaml_policies.py]: Oh yes, noted. Fixed it now.

**Review by gruebel — APPROVED** [review]
🏆

**Review by tronxd — APPROVED** [review]

**Pull request merged**